contracts/eoseosrateio/eoseosrateio.cpp

```diff
@@ -6,25 +6,53 @@
 using namespace eosio;
 using namespace std;
 
+namespace eosio {
+
+   constexpr name system_account{"eosio"_n};
+
+   struct voter_info {
+      name                owner;
+      name                proxy;
+      std::vector<name>   producers;
+      int64_t             staked = 0;
+      double              last_vote_weight = 0;
+      double              proxied_vote_weight= 0;
+      bool                is_proxy = 0;
+      uint32_t            flags1 = 0;
+      uint32_t            reserved2 = 0;
+      eosio::asset        reserved3;
+
+      uint64_t primary_key()const { return owner.value; }
+
+      EOSLIB_SERIALIZE( voter_info, (owner)(proxy)(producers)(staked)(last_vote_weight)(proxied_vote_weight)(is_proxy)(flags1)(reserved2)(reserved3) )
+   };
+
+   typedef eosio::multi_index< "voters"_n, voter_info >  voters_table;
+
+   bool is_proxy(name name) {
+      voters_table _voters(system_account, system_account.value);
+
+      auto it = _voters.find(name.value);
+      return it != _voters.end() && it->is_proxy;
+   }
+
+} /// namespace eosio
+
 CONTRACT eoseosrateio : public contract {
   public:
     using contract::contract;
 
     ACTION rateproducer(name user, name bp, string ratings_json) {
-      require_auth(user);
+      require_auth(_self);
```

Comment:
the user is the one that signs the rateproducer action not the the eoseosrateio contract